### ISSUE TITLE
EDID as control channel - possible?

### ISSUE THREAD

**Issue by showier-drastic** [steve-m/hsdaoh#3]
Hi,

I would like to ask whether it is possible to turn the EDID I2C into a downstream control channel? Maybe via some magic usb command to modify EDID table?

**Comment by steve-m** [comment]
Yes, that is possible. So far there was no real use for this, so I didn't really clean up the code and commit it. Did that now in 9b910ef.

Here is some example code I used for testing.

Host side, can be plugged into hsdaoh_test.c:
```c
int cnt = 1;
char message[256];

while (!do_exit) {
	sprintf(message, "Hello World via EDID #%d", cnt++);
	hsdaoh_write_edid_cmd_data(dev, (uint8_t *)message, strlen(message));
	usleep(100000);
}
```

Program running on a Raspberry Pi Pico connected via I2C to the MS2130:

```c
#include <stdio.h>
#include "pico/stdlib.h"
#include "hardware/i2c.h"

int main()
{
	stdio_init_all();

	i2c_init(i2c_default, 100 * 1000);
	gpio_set_function(PICO_DEFAULT_I2C_SDA_PIN, GPIO_FUNC_I2C);
	gpio_set_function(PICO_DEFAULT_I2C_SCL_PIN, GPIO_FUNC_I2C);
	gpio_pull_up(PICO_DEFAULT_I2C_SDA_PIN);
	gpio_pull_up(PICO_DEFAULT_I2C_SCL_PIN);

	int ret;
	uint8_t txdata = 0;
	uint8_t rxdata;
	uint8_t msglen;
	uint8_t seq = 0x01;
	uint8_t last_seq = 0;

	while (true) {
		/* read sequence number */
		txdata = 0;
		ret = i2c_write_blocking(i2c_default, 0x50, &txdata, 1, false);
		if (ret < 0)
			continue;

		ret = i2c_read_blocking(i2c_default, 0x50, &rxdata, 1, false);
		if (ret < 0)
			continue;

		if ((rxdata != 0xff) && (rxdata != 0x00)) {
			seq = rxdata;
			if (seq != last_seq) {
				/* read length field */
				txdata = 1;
				ret = i2c_write_blocking(i2c_default, 0x50, &txdata, 1, false);

				if (ret < 0)
					continue;


				ret = i2c_read_blocking(i2c_default, 0x50, &msglen, 1, false);

				if (ret < 0)
					continue;

				txdata = 2;
				ret = i2c_write_blocking(i2c_default, 0x50, &txdata, 1, false);
				if (ret < 0)
					continue;

				/* read actual message */
				char message[256];
				ret = i2c_read_blocking(i2c_default, 0x50, message, msglen, false);
				if (ret < 0)
					continue;

				message[msglen] = '\0';
				printf("New message with length %d: %s\n", msglen, message);

				last_seq = seq;
			}
		}
	}

	return 0;
}
```

**Comment by amalinowski75** [comment]
Hi @steve-m 
Have you ever tried to use hsdaoh_write_edid_cmd_data() with an actual EDID file? 
I tried today and I couldn't  make it work. The first problem is the size. All 1.3 and 1.4 EDIDs in binary format are 256 bytes long but the function expects it to be no longer than 254 bytes. I tried by skipping last 2 bytes but no luck.

And where did you get the information about the registers from?

**Comment by amalinowski75** [comment]
Jeez, I wasn't expecting so prompt response :) Thank you!
I did a quick test and it looks like it works. I'll play with it more tomorrow but the very first try is successful. 
Just had to change ```for (uint8_t i = 0; i < len; i++)``` to ```for (unsigned int i = 0; i < len; i++)```, ofc.

Thank you so much.

**Comment by amalinowski75** [comment]
I played with it and it looks like it works really good. Thank you again.
After uploading a new EDID the HDMI cable needs to be replugged to update the EDID on the connected computer.
Do you happen to know if there's a way to emulate HDMI disconnection on MS2131 side? It'd be nice not only to perform the EDID reload but also to disconnect the HDMI on demand.

**Comment by steve-m** [comment]
For that you need to toggle the hotplug detect (HPD) pin. It can be controlled via register 0xe003 bit 0 of the MS2130, so by writing 1 or 0 to the register you can generate a HPD pulse that should force the HDMI source to re-read the EDID.

**Comment by amalinowski75** [comment]
That seems to be more complicated.
Writing 0 or 1 to that register has no effect without updating EDID.

But after updating EDID writing 0 to 0xe003 indeed forces the source to reread EDID. Unfortunately only once. Next call to the sequence disconnects the source for good and I have to replug the grabber on the USB side. Then I can repeat the sequence and the first call works fine while the second one disconnects the source again. Replugging HDMI cable doesn't give any effect.

But when after EDID upload I write 0 and then 1 to the register it behaves the same way but then I can recover the HDMI connection by replugging the HDMI cable.

Any idea?

**Comment by amalinowski75** [comment]
I also tried to modify only the bit 0 so that I don't mess with other bits but nothing changed.

**Comment by steve-m** [comment]
Hm, I'd have to give it a try as well I guess.. in some test code I have (where I just tried toggling HPD) I also wrote 0xff to register 0xe000, but I don't remember what that was for. Maybe read that register when everything works and after using the sequece and compare the values.

**Comment by amalinowski75** [comment]
Hi @steve-m,
Sorry to bother you again but I am in need.
I noticed that after I upload EDID and I write anything (the value is not important) to the register 0xe003 the HPD pin goes down but I can't get it up. I noticed that to make things work again I need to disconnect +5V power (pin 18 of HDMI plug) and connect again.
Is there a way to do it in some programmatic way? Or maybe there is a way to trigger some kind of reset of MS2131 that will keep the modified EDID but restart the communication with the HDMI source?

**Comment by amalinowski75** [comment]
NVM, I manged to make everything work thanks to https://github.com/BertoldVdb/ms-tools/issues/21
